DOC-4519-audit-log (#117)
datastax/astra-streaming-docs: modules/operations/pages/monitoring/stream-audit-logs.adoc

= Astra Audit Logging

Stream audit logs through {product_name} to an external system.

This feature requires coordination with your DataStax representative.

To enable audit logging, you will need to provide the **Full Name** of your {product_name} topic and the streaming tenant's `client.conf` file to DataStax.

> == Create an {product_name} topic for audit logging

Author: Good point

If you already have an existing {product_name} topic you want to use for audit logging, skip to <<download-client-conf>>.

> . Log in to the Astra console using your credentials.
> . Select the appropriate organization for the streaming account you wish to use.
> . Click on **Streaming** from the left-hand menu.

Reviewer: ```suggestion
. In the {link-astra-portal}, go to **Streaming**.
```

> . Select an existing tenant or xref:astra-streaming:getting-started:index.adoc[Create a Streaming Tenant].

Reviewer: ```
. Select an existing tenant or xref:astra-streaming:getting-started:index.adoc#your-first-streaming-tenant[Create a Streaming Tenant] in AWS `us-east-2`.
. Add a xref:astra-streaming:getting-started:index.adoc#add-a-namespace-to-hold-topics[namespace] and xref:astra-streaming:getting-started:index.adoc#a-topic-to-roganize-messages[topic] to the tenant.
. On the *Namespace and Topics* page, click the new topic, and then copy the topic's **Full Name**, such as `persistent://aws-us-east-2-mk/*NAMESPACE_NAME*/*TOPIC_NAME*.
. If necessary, create additional audit log topics, and then record the **Full Name* for each topic.
```

You can combine or delete most of these steps.

The only unusual thing here is that the tenant must be in AWS us-east-2. Otherwise, you can save a lot of space by just linking to what you've already written.

I think it would be helpful to add why you would create multiple topics.

**The tenant must be located in AWS `useast2`.**
. Create a namespace in the tenant.
.. Select **Namespace and Topics** from the menu.
.. Click on **Create Namespace**.
.. Enter a name for your new namespace. In this example, we'll use `audit-logs`.
. Create a topic in the namespace.
.. Click **Add Topic** on the `audit-logs` namespace you just created.
.. Choose a name for your new topic. In this example, we'll use `dev`.
.. The default **Persistent** and **Non-partitioned** settings are fine.
.. Click on the new topic from the **Namespace and Topics** screen.
. Copy down the topic's **Full Name** value for your DataStax representative.
It will look like `persistent://aws-us-east-2-mk/audit-logs/dev`.
.. Set up as many additional topics as you need, and record the **Full Name** of each.

> [[download-client-conf]]
> == Download the client configuration file
> 

Reviewer: ```suggestion
```

> . From your streaming tenant, click on the **Connect** tab, and then click the **Download client.conf** button.
> . The `client.conf` file is downloaded to your local machine.

Reviewer: ```suggestion
. In the {link-astra-portal}, go to **Streaming**, and then click your audit log streaming tenant.
.  On the *Connect* tab, click **Download client.conf**.
```

> == Send information to DataStax
> 

Reviewer: ```suggestion
```

= Stream {astra_db} audit logs 

Stream your xref:astra-db-serverless:administration:view-account-audit-log.adoc[{astra_db} audit logs] through {product_name} to an external system.

> To enable audit log streaming, you must provide the **Full Name** of your {product_name} topic and the streaming tenant's `client.conf` file to {support_url}[{company} Support] or your account representative, or POST your configuration to the https://docs.datastax.com/en/astra-api-docs/_attachments/devops-api/index.html#tag/Organization-Operations/operation/configureTelemetry[Astra DevOps API telemetry endpoint].

Reviewer: ```suggestion
To enable audit log streaming, you must do one of the following:

 * Provide the **Full Name** of your {product_name} topic and the streaming tenant's `client.conf` file to {support_url}[{company} Support] or your account representative.
 * POST your configuration to the xref:astra-api-docs:ROOT:attachment$devops-api/index.html#tag/Organization-Operations/operation/configureTelemetry[Astra DevOps API telemetry endpoint].
```

Author: This is MUCH better, thank you

> == Create an {product_name} topic for audit logs

Reviewer: ```suggestion
== Create an {product_name} topic for audit logs

```

Audit log streaming requires a streaming tenant in the AWS `us-east-2` region.
You can create a new tenant with the xref:astra-streaming:getting-started:index.adoc[{product_name} quickstart] or use an existing {product_name} tenant.

. In the {link-astra-portal}, go to **Streaming**.
. Select an existing tenant or xref:astra-streaming:getting-started:index.adoc#your-first-streaming-tenant[Create a Streaming Tenant] in AWS `us-east-2`.
. Add a xref:astra-streaming:getting-started:index.adoc#add-a-namespace-to-hold-topics[namespace] and xref:astra-streaming:getting-started:index.adoc#a-topic-to-organize-messages[topic] to the tenant.
. On the *Namespace and Topics* page, click the new topic, and then copy the topic's **Full Name**, such as `persistent://aws-us-east-2-mk/*NAMESPACE_NAME*/*TOPIC_NAME*`.
. If necessary, create additional audit log topics, and then record the **Full Name** for each topic.

> Additional topics can help you organize audit logs by event type or other criteria.

Reviewer: ```suggestion
You can use topics to organize audit logs by event type or other criteria.
```

. In the {link-astra-portal}, go to **Streaming**, and then click your audit log streaming tenant.
. On the *Connect* tab, click **Download client.conf**.

> . Send your topic's full address and the `client.conf` file to your DataStax representative. DataStax will complete the setup.

Reviewer: ```suggestion
. To finalize the configuration, do one of the following:
+
* Send your topic's full name and the `client.conf` file to {support_url}[{company} Support] or your account representative, and then {company} will complete the setup.
* <<use-the-devops-api,Use the Astra DevOps API to complete the setup.>>
```

Author: Also much better

> == Configure the audit log with the DevOps API

Reviewer: ```suggestion
[#use-the-devops-api]
== Configure audit log streaming with the DevOps API
```

> You can configure audit log streaming using the Astra DevOps API instead of contacting your DataStax representative.
> 
> To do this, send a POST request to the Astra DevOps API telemetry endpoint. Include your topic’s full address and values from the `client.conf` file in the request. You'll need an **Organization Administrator** token to use this API.

Reviewer: ```suggestion
Ypu can use the xref:astra-api-docs:ROOT:attachment$devops-api/index.html#tag/Organization-Operations/operation/configureTelemetry[Astra DevOps API telemetry endpoint] to configure audit log streaming instead of providing the configuration details to {company} Support.

. In the {link-astra-portal}, create an application token with the **Organization Administrator** role, if you don't already have one.
```

> . Create the audit log streaming configuration.

Reviewer: ```suggestion
. To create the audit log streaming configuration, send a POST request with your topic’s full name and the required values from the tenant's `client.conf` file.
+
The `auth_strategy` and other authentication details depend on your Pulsar configuration.
```

+
[source,curl,subs="verbatim,quotes"]
----
curl --request POST \
  --url 'https://api.astra.datastax.com/v2/organizations/**ORG_ID**/telemetry/auditLogs' \
  --header 'Accept: application/json' \
  --header 'Authorization: Bearer **APPLICATION_TOKEN**' \
  --data '{
  "pulsar": {
    "endpoint": "pulsar+**BROKER_SERVICE_URL**",
    "auth_strategy": "token",
    "topic": "**TOPIC_FULL_NAME**",
    "auth_name": "token",
    "token": "**PULSAR_AUTHENTICATION_TOKEN**"
  }
}'
----
+

> Response:
> +
> [source,plain]
> ----
> HTTP/1.1 202 Accepted
> ----

Reviewer: ```suggestion
.Response
[%collapsible]
====
[source,plain]
----
HTTP/1.1 202 Accepted
----
====
```

> +

Reviewer: ```suggestion

```

> . Retrieve the audit log streaming configuration.

Reviewer: ```suggestion
. To retrieve and verify the audit log streaming configuration, send a GET request:
```

+
[source,curl,subs="verbatim,quotes"]
----

> curl --request GET \
>   --url 'https://api.astra.datastax.com/v2/organizations/**ORG_ID**/telemetry/auditLogs' \
>   --header 'Accept: application/json' \
>   --header 'Authorization: Bearer **APPLICATION_TOKEN**' \
>   --data '{
>   "pulsar": {
>     "endpoint": "pulsar+**BROKER_SERVICE_URL**",
>     "auth_strategy": "token",
>     "topic": "**TOPIC_FULL_NAME**",
>     "auth_name": "token",
>     "token": "**PULSAR_AUTHENTICATION_TOKEN**"
>   }
> }'

Reviewer: ```suggestion
curl --request GET \
  --url 'https://api.astra.datastax.com/v2/organizations/**ORG_ID**/telemetry/auditLogs' \
  --header 'Authorization: Bearer **APPLICATION_TOKEN**'
```

Reviewer: GET does not have a body.

https://docs.datastax.com/en/astra-api-docs/_attachments/devops-api/index.html?_gl=1*5e1j5c*_gcl_au*MjgzNDY4MjI0LjE3MjIyODk4NTM.#tag/Organization-Operations/operation/getTelemetryConfig

Author: Confirmed
{"pulsar":{"endpoint":"pulsar+ssl://pulsar-aws-useast2.streaming.datastax.com:6651","topic":"persistent://aws-us-east-2-mk-2/default/audit-log","auth_strategy":"token","token":"********","auth_name":"token"}}%

Reviewer: Confirmed that it does or does not have a body? I don't see a request body in the reference, just a response example.

----
+

> Response:
> +
> [source,json]
> ----
> {
>   "pulsar": {
>     "endpoint": "pulsar+ssl://pulsar-aws-useast2.streaming.datastax.com:6651",
>     "topic": "persistent://aws-us-east-2-mk-2/default/audit-log",
>     "auth_strategy": "token",
>     "token": "********",
>     "auth_name": "token"
>   }
> }
> ----

Reviewer: ```suggestion
.Response
[%collapsible]
====
[source,json]
----
{
  "pulsar": {
    "endpoint": "pulsar+ssl://pulsar-aws-useast2.streaming.datastax.com:6651",
    "topic": "persistent://aws-us-east-2-mk-2/default/audit-log",
    "auth_strategy": "token",
    "token": "********",
    "auth_name": "token"
  }
}
----
====
```

> 

Reviewer: ```suggestion

. To delete an audit log streaming configuration, xref:astra-api-docs:ROOT:attachment$devops-api/index.html#tag/Organization-Operations/operation/deleteTelemetryConfig[send a DELETE request].
```
